design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,510 @@
+  Furthermore, it seems feasible to generate this boilerplate code using macros
+  (the `run()` method, the handler `ActorAHandler` struct and implementation, the `AMessage` enum).
+* The handler type is a great place to improve both code readability and code flexibility
+  (A wrapping similar to what has been done for the third variant can be done for the first to hide lock acquisition).
+  We will use such handlers and adapters to encapsulate messages related features
+  as sending high priority messages or sending a request for which a response is awaited.
+* Similarly, the channel receiver can be encapsulated to add features
+  like reading high priority messages first or awaiting a response
+* Key points are not addressed by this example.
+  * Who creates the actor channels and instances?
+  * How actors and peers are connected to each others?
+  * Who spawns the actor `run()` method?
+
+
+
+### Messages
+
+Messages must flow freely between actors with no constraints on ownership and thread.
+As they are used to improved observability, they must be `Display`.
+
+```rust
+/// A message exchanged between two actors
+pub trait Message: Clone + Display + Send + Sync + 'static {}
+
+/// There is no need to tag messages as such
+impl<T: Clone + Display + Send + Sync + 'static> Message for T {}
+```
+
+Typical examples of thin-edge messages are telemetry data, operation requests and outcomes,
+but also low level messages as MQTT messages, HTTP requests and responses,
+and even system specific messages as file-system events and update requests.
+
+To be discussed:
+* Do the messages need to be `Clone`?
+  This has to be considered along the idea of using `oneshot` channel for the response to a request.
+  It might be better to be explicit
+  and use `Message + Clone` in contexts where messages are broadcast.
+
+### Channels
+
+Multi-producer single-consumer (`mpsc`) channels are used to exchange messages between actors.
+
+* A channel is created for each actor instance.
+* The receiver of this channel is given to the recipient actor.
+* Clones of the channel sender are given to any actor that needs to send messages to this instance.
+* With this setup, each actor instance owns
+  - an `mpsc::Receiver` end
+  - and a bunch of `mpsc::Sender`s, one per peer. 
+* The actors process then in turn the received messages,
+  updating their internal state and sending messages to their peers. 
+
+Having a single receiver per actor improves modularity, observability and testability,
+since all the inputs for an actor are going through this receiver.
+Similarly, having the peers of an actor materialized by channel senders
+helps to understand and test the actors in isolation.
+
+However, for this to work, several points need to be addressed, regarding:
+
+* Message types
+* Channel types
+* Channel creation and ownership
+* Actor with no inputs
+* Addressing responses
+* Out-of-band runtime messages
+* Size of the channel buffers
```

Comment:
```suggestion
* Size of the channel buffers
* Persistence of the channels
```
For robustness and resumability.

Reply:
Indeed a key point. However, I will put that in another place in the document - as a refinement.

Reply:
This has been added as a "nice to have".